[Security Solution] Add `ComparisonSide` component

nikitaindik opened the pull request:
**Partially addresses: https://github.com/elastic/kibana/issues/171520**

## Summary

This PR adds the `ComparisonSide` component for the ThreeWayDiff UI ([see it on the Miro diagram](https://miro.com/app/board/uXjVK0gqjjQ=/?moveToWidget=3458764594147853908&cot=14)). 

`ComparisonSide` lets the user compare field values from the two selected rule versions. It will be displayed on the left side of the upgrade flyout.  

You can view and test it in Storybook by running `yarn storybook security_solution` in the root Kibana dir. Go to `http://localhost:9001` once the Storybook is up and running.

https://github.com/user-attachments/assets/e71ae626-d0f7-43ae-8324-f3d4ea540b02


Also updated `react-diff-view` to the latest version (`3.2.0` -> `3.2.1`)

nikitaindik commented:
/ci

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

nikitaindik commented:
/ci

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/subfield_header.tsx:
```diff
@@ -0,0 +1,50 @@
+}
+
+export function SubfieldHeader({ fieldName, subfieldName }: SubfieldHeaderProps) {
+  if (!shouldDisplaySubfieldLabelsForField(fieldName)) {
```
`SubfieldHeader` as a leaf component shouldn't know about implementation details of the other components. This logic should be a few levels up in `Subfield` component. It's also good to leave a comment why header isn't always shown.

nikitaindik replied:
Done.

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/subfield_header.tsx:
```diff
@@ -0,0 +1,50 @@
+
+function shouldDisplaySubfieldLabelsForField(fieldName: string): boolean {
+  return [
+    'data_source',
```
Hardcoded field names make it harder to maintain this functionality later on when we decide to remove or rename some fields. The good way is to specify a union of rule field names.

nikitaindik replied:
Thanks! Updated.

maximpn reviewed (commented):
@nikitaindik thanks for you effort in implementing `ComparisonSide` component. It looks nice and works well. I tested in storybooks.

There are no tests for the added component here. Do you plan to add them in a separate PR? 

I left comments regarding subfield changes implementation here. Feel free to reach me for additional clarification.

nikitaindik commented on package.json:
```diff
@@ -1148,7 +1148,7 @@
     "react": "^17.0.2",
     "react-ace": "^7.0.5",
-    "react-diff-view": "^3.2.0",
+    "react-diff-view": "^3.2.1",
```
It fixes a minor bug with displaying newlines at the end of the diff: https://github.com/otakustay/react-diff-view/blob/master/CHANGELOG.md#321-2024-02-18 Not a big deal for us, but I decide to update anyways.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/json_diff/diff_view.tsx:
```diff
@@ -111,15 +112,32 @@ const renderGutter: RenderGutter = ({ change }) => {
+ * @returns {string} A unified diff string representing the changes.
+ */
+const convertChangesToUnifiedDiffString = (changes: Change[]) => {
+  const unifiedDiff: string = unidiff.formatLines(changes, {
```
Underrefactored a bit. Adding an explicit return type is a good idea! Added.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/get_subfield_changes_for_field.ts:
I have thought about it but wasn't sure if splitting it was worth the effort. Now since you brought that up I have refactored it into multiple files. 👍

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/comparison_side.stories.tsx:
```diff
@@ -0,0 +1,315 @@
+  diff_outcome: ThreeWayDiffOutcome.StockValueCanUpdate,
+  merge_outcome: ThreeWayMergeOutcome.Target,
+  has_update: true,
+  conflict: ThreeWayDiffConflict.NONE,
```
Actually my main idea was to just see / show how different field types (single / multi-line, numbers, arrays) are rendered. I didn't pay much attention to `conflict` since it doesn't directly affect this component.

nikitaindik replied:
> Input data should be accurate.

I agree. And I think it is accurate in this case. 

`conflict` is `NONE`, because it's an `AAB` situation, so the target value is taken and there's no conflict. Isn't it?

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/comparison_side.stories.tsx:
```diff
@@ -0,0 +1,315 @@
+    diff_outcome: ThreeWayDiffOutcome.CustomizedValueCanUpdate,
+    merge_outcome: ThreeWayMergeOutcome.Merged,
+    has_update: true,
+    conflict: ThreeWayDiffConflict.SOLVABLE,
```
This one is solvable. I grabbed it from our [unit tests](https://github.com/nikitaindik/kibana/blob/23c709547e132bbc753908ce41c62a01ed37a4ea/x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/multi_line_string_diff_algorithm.test.ts#L95).

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/comparison_side.stories.tsx:
```diff
@@ -0,0 +1,315 @@
+    diff_outcome: ThreeWayDiffOutcome.CustomizedValueCanUpdate,
+    merge_outcome: ThreeWayMergeOutcome.Merged,
+    has_update: true,
+    conflict: ThreeWayDiffConflict.SOLVABLE,
```
With this one I wanted to check / showcase a real-life situation. A few changes in a setup guide - a few words here, a sentence there... What do you think about have 1 or 2 longer mocks to check a more real-life case?

maximpn replied:
I agree it's beneficial during development phase in storybook to verify the component works with real multiline data. Though preparing and maintaining such data is tricky since spotting differences manually takes time. Let's use simplified abstract text later for testing and leave current text here unchanged.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/get_subfield_changes.ts:
```diff
@@ -0,0 +1,123 @@
+  switch (fieldName) {
+    case 'data_source':
+      return getSubfieldChangesForDataSource(
+        oldFieldValue as RuleDataSource | undefined,
+        newFieldValue as RuleDataSource | undefined
```
Def agree this isn't the best approach. I couldn't get it to work quickly without typecasting. I would appreciate it if we could pair and improve the types here.

nikitaindik replied:
Discussed this offline. It turned out it's not possible to avoid typecasting since TS doesn't support narrowing union types. Added an explainer comment in code. We may want to discuss the type situation at the upcoming Tech Time.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/json_diff/diff_view.tsx:
```diff
@@ -154,18 +170,31 @@ const CustomStyles: FC<PropsWithChildren<unknown>> = ({ children }) => {
       width: ${euiTheme.size.xl};
     }
 
+    /* Hide the redundant second gutter column in "unified" view */
```
Updated the comment.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/comparison_side.stories.tsx:
Discussed this offline. Decided to leave the Storybook mocks as they are for now because they help to see how the component would look with a "real-life like" input.

Will definitely use more basic mocks for unit tests.

banderror replied:
@maximpn I agree with @nikitaindik, Storybook allows us to see how a component will look like in the app before we integrate it with it. It's our chance to use close-to-real input data to see how it will look in reality.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/get_subfield_changes_for_field.ts:
```diff
@@ -0,0 +1,456 @@
+} from '../../../../../../../common/api/detection_engine';
+import type { SubfieldChange } from './types';
+
+export const sortAndStringifyJson = (fieldValue: unknown): string => {
```
Discussed this offline. Renamed to `stringifyToSortedJson`.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/get_subfield_changes_for_field.ts:
```diff
@@ -0,0 +1,456 @@
+
+  const hasTypeChanged = oldType !== newType;
+
+  return [
+    ...(hasTypeChanged
+      ? [
+          {
+            subfieldName: 'type',
+            oldSubfieldValue: oldType,
+            newSubfieldValue: newType,
+          },
+        ]
+      : []),
+    ...(oldIndexPatterns !== newIndexPatterns
+      ? [
+          {
+            subfieldName: 'index_patterns',
+            oldSubfieldValue: oldIndexPatterns,
+            newSubfieldValue: newIndexPatterns,
+          },
+        ]
+      : []),
+    ...(oldDataViewId !== newDataViewId
+      ? [
+          {
+            subfieldName: 'data_view_id',
+            oldSubfieldValue: oldDataViewId,
+            newSubfieldValue: newDataViewId,
+          },
+        ]
+      : []),
```
Good one. Thanks! Updated.

nikitaindik commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/get_subfield_changes_for_field.ts:
```diff
@@ -0,0 +1,456 @@
+  return stringify(fieldValue, { space: 2 });
+};
+
+export const getSubfieldChangesForDataSource = (
```
Discussed this offline. Only one arg can be `undefined` but describing it with TS is difficult in this situation.

nikitaindik commented:
/ci

nikitaindik commented:
Thanks for your feedback @maximpn! I have addressed your comments. Feel free to take another look!

As for the tests, I plan to add them later once the base functionality is completed and the test plan is ready.

dplumlee commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/utils.ts:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Stringifies a field value to an alphabetically sorted JSON string.
+ */
+export const stringifyToSortedJson = (fieldValue: unknown): string => {
+  if (fieldValue === undefined) {
+    return '';
+  }
+
+  if (typeof fieldValue === 'string') {
+    return fieldValue;
+  }
+
+  return stringify(fieldValue, { space: 2 });
+};
```
nit: we define an identical function for the per-field diff [here](https://github.com/elastic/kibana/blob/5f49cc50178d4abc67be420534f8cfeafb60c954/x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/per_field_diff/get_field_diffs_for_grouped_fields.ts#L18), might be worth abstracting both of them out

nikitaindik replied:
Yes, it's actually the same function. My idea was to just copy-paste it from the per-field diff, since we are going to remove the per-field diff tab and substitute it with this new tab once it's ready. But now I think it might be a good idea to abstract it away in a shared dir.

nikitaindik replied:
Yes, it's actually the same function. I copy-pasted it from the per-field diff. I figured that copy-pasting is fine since we are going to remove the per-field diff code at some point once this new tab is ready. But since the function is very generic, perhaps it's a good idea to extract it away.

Any recommendations for a good place to put it? I couldn't find a suitable utils file. If you have a recommendation, I can move it there in the next PR.

dplumlee replied:
Not exactly the same thing but we could put it in this helpers file: `x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/helpers.ts`? I probably should've put it there initially as we have another similar function in `rule_diff_tab.tsx` that you originally added.

nikitaindik replied:
Thanks! Will add to my todo for the next PR.

dplumlee commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/subfield.tsx:
```diff
@@ -0,0 +1,50 @@
+import { DiffView } from '../../json_diff/diff_view';
+import { SubfieldHeader } from './subfield_header';
+
+const FIELDS_WITH_SUBFIELDS: Array<keyof DiffableAllFields> = [
+  'data_source',
+  'kql_query',
+  'eql_query',
+  'esql_query',
+  'threat_query',
+  'rule_schedule',
+  'rule_name_override',
+  'timestamp_override',
+  'timeline_template',
+  'building_block',
+  'threshold',
+];
```
nit: this could be put in a `constants.ts` file in this folder or in the above folder, we'll likely have other stuff to add to it or reference these at some point in the future (tests, etc)

nikitaindik replied:
Done!

dplumlee reviewed (approved):
Component looks great @nikitaindik! This was my first time diving deep with storybooks, very cool. Left a few nits related to overall code structure but I think the refactored code looks good!

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/get_subfield_changes/index.ts:
```diff
@@ -0,0 +1,109 @@
+  newFieldValue?: DiffableAllFields[FieldName]
+): SubfieldChange[] => {
+  switch (fieldName) {
+    /* Typecasting here because narrowing a union type (`DiffableAllFields[FieldName]`) is not supported by TS */
```
```suggestion
    /* Typecasting `oldFieldValue` and `newFieldValue` to corresponding field type `DiffableAllFields[*]` is required here since `oldFieldValue` and `newFieldValue` concrete types depend on `fieldName` but TS isn't track that. */
```

maximpn commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/three_way_diff/comparison_side/types.ts:
```diff
@@ -0,0 +1,12 @@
+ * 2.0.
+ */
+
+export interface SubfieldChange {
```
We discussed that it's good to define `SubfieldChanges` type as well looking like `type SubfieldChanges = Readonly< SubfieldChange[]>;`. Are there issues with that?

maximpn reviewed (approved):
@nikitaindik thank you for addressing my comments 👍 

I left a few more but these don't look critical. @dplumlee mentioned that a function producing sorted JSON already exists. It worths to reduce code duplication.
